src/bindings/jna/libelektra5j/src/main/java/org/libelektra/KDB.java

```diff
@@ -90,6 +103,49 @@ public void close (final Key parentKey)
 		Elektra.INSTANCE.kdbClose (kdb, parentKey.get ());
 	}
 
+	/**
+	 * Creates a contract for use with {@link KDB#open} that mounts and configures the gopts plugin
+	 *
+	 * @param contract    the KeySet into which the contract is written
+	 * @param args        the arguments that will be converted into argc and argv for gopts
+	 * @param env         the environment variables that gopts will use
+	 * @param parentKey   the parent key that gopts will use
+	 * @param goptsConfig the config KeySet used for mounting gopts
+	 *
+	 * @throws IllegalArgumentException if any of the arguments are null
+	 */
+	public static void goptsContract (final KeySet contract, final String[] args, final String[] env, final Key parentKey,
```

Comment:
It is kind of related in the sense that the only reason to call this function is that you need a contract to `KDB.open`.

Reply:
> Btw. I think the G inside is also confusing, is this for global plugin?

I have no idea, why I chose `g`. I just now the plugin can't be called `opts` because that would collide with the `opts` library.

Reply:
> With that logic the C API should be also kdbOptsContract?

Not really, in C we always use the `elektra` prefix for helper functions and `kdb`/`key`/`ks` for core API only.

I can move the function to it's own class, but I think the discussion is a bit pointless, since AFAIK the Java API will be rewritten soon anyway.

Reply:
It is about how the contracts should be modularized conceptually, we cannot delegate this to individual binding writers. All bindings should somehow resemble a similar structure.

Reply:
> All bindings should somehow resemble a similar structure.

Absolutely not. Bindings should be designed to make them feel natural to the language they are written in. The binding writer should absolutely decide what is the most idiomatic approach for the language.

I could for example also imagine an approach like this for Java:

```java
KeySet contract = new ContractBuilder().gopts (...).io(...).notifications(...).foo(...).build();
KDB.open (contract, parentKey);
```

In Kotlin you might wanna do:

```kotlin
KeySet contract = contractBuilder {
  gopts(...)
  io(...)
  notifications(...)
  foo(...)
}
```

or maybe

```kotlin
KeySet contract = contractBuilder {
  gopts {
    ...
  }
  foo {
    ...
  }
}
```

In Javascript it might just be:

```js
const constract = buildContract ({
  gopts: { ... },
  io: { ... },
  foo { ... },
});
```

Every language is different and that is why the API for every language should be different.

Reply:
You are misinterpreting. I was talking about the structure, like the 3 classes. Now we need to find a solution where the contract fits in. Your examples indicate you see the contract (builder) as something separated from the 3 classes, which is exactly what my initial comment was about.

In this case (because changes in the Java API are planned anyway) we can also simply open an issue clearly saying that having the goptsContract in KDB was a hack and is not intended to be released that way.

Reply:
> You are misinterpreting. I was talking about the structure [...] Your examples indicate you see the contract (builder) as something separated

The structure is also determined by the language. Some languages allow functions only inside objects/classes others allow global functions. Some have namespaces others don't. Sometimes you can use a single function other times multiple functions are better.

If you want to know about structure regarding my examples: I didn't really think about it when writing them, but here we go:

In Java, I would either call the class `KdbContractBuilder` and or make `ContractBuilder` an inner class of `KDB`.
In Kotlin, I would probably put the `contractBuilder` function directly into the package `org.libelektra.kdb`, together with the `KDB` class.
In Javascript, I would similarly but the function `buildContract` into a `kdb` module, together with the `KDB` class. But depending on how you import that module, you might end up calling `kdb.buildContract`.

Reply:
> In Java, I would either call the class KdbContractBuilder and or make ContractBuilder an inner class of KDB.

Ok, then we can let it be like it is now, the function to the inner class can easily be changed later.